Accounting support

MirageTurtle opened the pull request:
Add basic logging and accounting support via hourly parsing logs.

MirageTurtle committed:
Add access log support.

Just support access log for accounting.

MirageTurtle committed:
Add `ganted2radius` to parse and send accounting data to radius.

Use crond to do it hourly.

MirageTurtle committed:
Use github.com/robfig/cron instead of crond.

iBug commented on src/main.go:
```diff
@@ -116,6 +119,20 @@ func (r *RadiusCredentials) StartGCWorker() {
 }
 
+func (r *RadiusCredentials) accountingCron() *cron.Cron {
+        // hourly accounting cron job
```
This is indented with space. Consider adding a save hook that calls `gofmt` (or [`goimports`](https://pkg.go.dev/golang.org/x/tools/cmd/goimports)) for your editor.

MirageTurtle replied:
Fixed.

iBug commented on src/main.go:
```diff
@@ -116,6 +119,20 @@ func (r *RadiusCredentials) StartGCWorker() {
+func (r *RadiusCredentials) accountingCron() *cron.Cron {
+        // hourly accounting cron job
+     	c := cron.New()
+	c.AddFunc("0 0 * * * *", func() {
```
Is this really "hourly"?

MirageTurtle replied:
Go cron uses 6 space-separated fields, with second filed. Use `@hourly` if you prefer.

iBug commented on src/main.go:
```diff
@@ -116,6 +119,20 @@ func (r *RadiusCredentials) StartGCWorker() {
+     	c := cron.New()
+	c.AddFunc("0 0 * * * *", func() {
+		// accounting
+		err := r.accounting("/var/log/ganted/access.log", "/var/log/ganted/archive.log")
```
Shouldn't this use file paths from environment instead of hard-coded paths?

MirageTurtle replied:
Fixed, use environment variable `GANTED_LOG_DIR`, but force the `access.log`, `error.log`, `archive.log` are under `GANTED_LOG_DIR`.

iBug commented on src/main.go:
```diff
@@ -131,10 +148,32 @@ func init() {
 
+func init() {
+	// init dir /var/log/ganted
+	if _, err := os.Stat("/var/log/ganted"); os.IsNotExist(err) {
```
Maybe put the log directory in environment variables.

MirageTurtle replied:
Using environment variable `GANTED_LOG_DIR`.

iBug commented on src/main.go:
```diff
@@ -157,18 +196,32 @@ func main() {
 		},
 	}
 	credentials.StartGCWorker()
+	c :=credentials.accountingCron()
```
`gofmt` or `goimports` should fix this.

MirageTurtle replied:
Fixed

MirageTurtle committed:
Remove hard-coded paths, and go fmt.

taoky commented on src/go.sum:
```diff
@@ -1,5 +1,7 @@
 github.com/kisom/netallow v0.0.0-20200609175051-08f6b004e41a h1:4T7cUpk4OIqaxn7i41yVYEU7/4gjTKuvqlrSqbBwwe0=
 github.com/kisom/netallow v0.0.0-20200609175051-08f6b004e41a/go.mod h1:fHrMiR3Isu09r3MBg9oqlp0E+qBtRp6qsig0hpmgXYg=
+github.com/robfig/cron v1.2.0 h1:ZjScXvvxeQ63Dbyxy76Fj3AT3Ut0aKsyd2/tl3DTMuQ=
```
It looks like upstream has published robfig/cron v3: <https://github.com/robfig/cron>

Are there any reasons to stick to v1.2?

MirageTurtle replied:
I just `go get` without specific version and it gets v1.2. I will upgrade it to v3 later.

MirageTurtle replied:
Upgrade at new commit.

iBug commented on src/go-socks5/socks5.go:
```diff
@@ -161,11 +198,22 @@ func (s *Server) ServeConn(conn net.Conn) error {
+	if err := s.handleRequest(request, wrappedConn); err != nil {
 		err = fmt.Errorf("Failed to handle request: %v", err)
 		s.config.Logger.Printf("[ERR] socks %s: %v", remoteAddr, err)
 		return err
```
We should continue to log the connection even if an error happens mid-way.

MirageTurtle replied:
defer the log writing at new commit.

taoky commented on src/ganted2radius.go:
```diff
@@ -0,0 +1,138 @@
+	}
+	defer srcFile.Close()
+
+	dstFile, err := os.Create(dst)
```
If we're logging for auditing reasons, wouldn't this overwrite existing log?

MirageTurtle replied:
Write some logic to rotate log files? And archive daily? like https://github.com/ustclug/light-server/blob/master/.docker/rootfs/etc/periodic/hourly/squid2radius#L6-L10

taoky replied:
Maybe we could implement with this logic:

0. Compress all existing access-\<datetime>.log to access-\<datetime>.log.zst
1. Rename access.log to access-\<datetime>.log
2. Ask access log logger to reopen access.log, which would let it just stop writing to the renamed access log
3. Process access-\<datetime>.log

MirageTurtle replied:
Implemented at new commit.

MirageTurtle committed:
Upgrade cron to v3, defer write the access log, rotate the log files

MirageTurtle commented on src/ganted2radius.go:
```diff
@@ -0,0 +1,198 @@
+func archiveLogs(logDir string) error {
+	logPattern := regexp.MustCompile(`^access-\d{14}\.log$`)
+	// compress all access-<datetime>.log files to access-<datetime>.log.zst
+	err := filepath.Walk(logDir, func(path string, info os.FileInfo, err error) error {
```
I'm a little concerned about the low performance of this operation when the number of logs is relatively large.
The solutions I can think of:
1) List all files and filter them first. But it seems that Go doesn't have functions to do this directly, or I miss something.
2) Use goroutine to speed up this.
3) Assume there is only one `access-<datetime>.log` that needs to be compressed.

taoky replied:
Emmm... Is `filepath.Walk()` slow? Maybe we could use `filepath.WalkDir()` and check filename first to avoid unnecessary `lstat()`.

Also I think that the compress part could be put into a goroutine (if it is slow), as we don't need to wait for compression to be done before accounting. Another approach is to do rotate only in the go program, and if there're already 24 files (hours) inside we call an external script to do extra logrotating and compressing like https://github.com/ustclug/light-server/blob/e1aa99d4790d4dd163fbe8771aa3093d40078548/.docker/rootfs/etc/periodic/hourly/squid2radius#L7-L9.

MirageTurtle replied:
Resolved at new commit. Just rotate when accounting, and archive(concatenate and compress) every 24 log files (daily).

MirageTurtle committed:
Just rotate when accounting, and archive every 24 logfiles(daily)

taoky commented on src/ganted2radius.go:
```diff
@@ -0,0 +1,249 @@
+			// Else, everything is ok, delete the original log files
+			if _, err := os.Stat(archiveFilePath); !os.IsNotExist(err) {
+				if err := os.Remove(archiveFilePath); err != nil {
+					fmt.Errorf("err when removing file %s", archiveFilePath)
```
`fmt.Errorf` returns an `error` instead of printing it out.

taoky commented on src/ganted2radius.go:
```diff
@@ -0,0 +1,249 @@
+		logFiles = logFiles[:maxBackup-1]
+		// create archive log file with current date
+		date := time.Now().Format("20060102")
+		archiveFileName := fmt.Sprintf("access-%s.log", date)
```
Use a different name like `archived-access-%s.log`?

MirageTurtle committed:
rename archive file, make sure err during removing could be handled

Delete `logFiles = logFiles[:maxBackup-1]`

iBug commented on src/ganted2radius.go:
```diff
@@ -0,0 +1,246 @@
+		}
+		// compress access-<date>.log
+		if err := compressFile(archiveFilePath); err != nil {
+			return err
```
`archiveFile` may remain open at this point. Better use `defer archiveFile.Close()` right after successfully opening.

iBug commented on src/ganted2radius.go:
```diff
@@ -0,0 +1,246 @@
+			}
+			_, err = io.Copy(archiveFile, src)
+			if err != nil {
+				return err
```
Same as below, `src` may go unclosed.

iBug commented on src/main.go:
```diff
@@ -131,10 +153,41 @@ func init() {
+	// init dir GANTED_LOG_DIR
+	gantedLogDir := getEnv("GANTED_LOG_DIR", "/var/log/ganted")
+	if _, err := os.Stat(gantedLogDir); os.IsNotExist(err) {
+		os.Mkdir(gantedLogDir, 0755)
```
Shall we `panic()` if the log directory doesn't exist and can't be created either?

iBug commented on src/main.go:
```diff
@@ -131,10 +153,41 @@ func init() {
+		log.Fatalf("[ERR] Open log file: %s", err)
+		return err
+	}
+	logger.SetOutput(file)
```
Consider fetching the previous writer and close it if valid. Example: [iBug/fakessh](https://github.com/iBug/fakessh/blob/8d175d6cc1e002a6e921059a98e676272254bc64/main.go#L48)

iBug reviewed (changes requested):
There are many occurrences of `fmt.Printf` and `fmt.Fprintf(os.Stderr`, they're better replaced by `log.Printf` instead.

MirageTurtle committed:
Requested changes of https://github.com/ustclug/light-socks5/pull/3

1. defer the close of `archiveFile` and `src`
2. use `log.Printf` to replace `fmt.Printf` and `fmt.Fprintf`
3. `panic()` if the log directory does not exist and can not be
created either
4. fetching the previous writer and close it if valid

iBug commented on src/main.go:
```diff
@@ -157,20 +156,25 @@ func init() {
 	if _, err := os.Stat(gantedLogDir); os.IsNotExist(err) {
-		os.Mkdir(gantedLogDir, 0755)
+		if err := os.Mkdir(gantedLogDir, 0755); err != nil {
+			panic(err)
```
`log.Fatalf` should do better here. It prints a formatted message and exits the program.

MirageTurtle replied:
I considered it, but I miss it because I think err will contain enough information. But It would be better that we explicitly print message. I fixed it.

MirageTurtle committed:
Format error message, use `log.Fatalf` for panic message of `Mkdir`

iBug commented on src/ganted2radius.go:
```diff
@@ -0,0 +1,247 @@
+		return err
+	}
+	defer file.Close()
+	// create the compressed file, err if it already exists
```
Code does not match comment. [`os.Create`](https://pkg.go.dev/os#Create) says:

> Create creates or truncates the named file. If the file already exists, it is truncated.

MirageTurtle committed:
fix: the issue of ensuring file existence and related error return.

add some comments, and fix some error message typo.

iBug commented on src/ganted2radius.go:
```diff
@@ -82,9 +93,12 @@ func archiveLogs(logDir string, maxBackup int) error {
+		// check if access-<date>.log.zst exists first
+		if _, err := os.Stat(archiveFilePath + ".zst"); err == nil {
 			return fmt.Errorf("file %s exists", archiveFileName+".zst")
+		} else if !os.IsNotExist(err) {
```
From [`os.IsNotExist`](https://pkg.go.dev/os#IsNotExist):

> This function predates [errors.Is](https://pkg.go.dev/errors#Is). It only supports errors returned by the `os` package. New code should use `errors.Is(err, fs.ErrNotExist)`.

Note that there's a second occurrence above that should be fixed as well.

iBug reviewed (approved):
One last minor issue to fix, so I'm giving this a preemptive approve (i.e. I'm fine even it's left alone).

MirageTurtle committed:
chore: `os.IsNotExist(err)` to `errors.Is(err, os.ErrNotExist)`

https://github.com/ustclug/light-socks5/pull/3/files/ff51f6867bce2d7a48663d874f1c2b359264b98b..3c95040132b75d0fc3db01eaf80b61ea320d6c50#r1938473747
